lib/src/Base/Geom/IntervalMesher.cxx

```diff
@@ -133,21 +134,51 @@ Mesh IntervalMesher::build(const Interval & interval) const
     } // j
     // Second the simplices
     Mesh::IndicesCollection simplices(0, Indices(3));
-    UnsignedInteger vertexIndex(0);
+    UnsignedInteger cellIndex(0);
     Indices index(3);
     for (UnsignedInteger j = 0; j < n; ++j)
     {
       for (UnsignedInteger i = 0; i < m; ++i)
       {
-        index[0] = vertexIndex;
-        index[1] = vertexIndex + 1;
-        index[2] = vertexIndex + 1 + m;
-        simplices.add(index);
-        index[0] = vertexIndex + 2 + m;
-        simplices.add(index);
-        ++vertexIndex;
+        // The current cell is
+        // c--d
+        // |  |
+        // a--b
+        const UnsignedInteger a(cellIndex);
+        const UnsignedInteger b(cellIndex + 1);
+        const UnsignedInteger c(cellIndex + 1 + m);
+        const UnsignedInteger d(cellIndex + 2 + m);
+        if (diamond)
+        {
+          const NumericalPoint center((vertices[a] + vertices[b] + vertices[c] + vertices[d]) * 0.25);
+          const UnsignedInteger centerIndex(vertices.getSize());
+          vertices.add(center);
+          index[0] = a;
+          index[1] = b;
+          index[2] = centerIndex;
+          simplices.add(index);
+          index[0] = b;
+          index[1] = d;
+          simplices.add(index);
+          index[0] = d;
+          index[1] = c;
+          simplices.add(index);
+          index[0] = c;
+          index[1] = a;
+          simplices.add(index);
+        }
+        else
+        {
+          index[0] = a;
+          index[1] = b;
+          index[2] = c;
+          simplices.add(index);
+          index[0] = d;
```

Comment:
IMHO [d,c,b] is better than [d,b,c] because normals are consistent.  This has no impact on your use case, but this may hurt other applications.  I did not check if this applies to 3D too.